### ISSUE TITLE
Wrong detect of `Parsing error: invalid-first-character-of-tag-name` in expression.

### ISSUE THREAD

**Issue by tinymins** [vuejs/eslint-plugin-vue#370]
**Tell us about your environment**

* **ESLint Version:**

4.16.0

* **eslint-plugin-vue Version:**

4.2.2

* **Node Version:**

v8.1.3

**Please show your full configuration:**

```
const isDev = process.env.NODE_ENV === 'development';

// http://eslint.org/docs/user-guide/configuring
module.exports = {
  root: true,
  parserOptions: {
    parser: 'babel-eslint',
    sourceType: 'module'
  },
  env: {
    browser: true,
  },
  extends: [
    'eslint:recommended',
    'plugin:vue/recommended', // or 'plugin:vue/base'
    'airbnb-base',
  ],
  // required to lint *.vue files
  plugins: [
    'vue',
  ],
  // check if imports actually resolve
  'settings': {
    'import/resolver': {
      'webpack': {
        'config': 'build/webpack.base.conf.js'
      }
    },
  },
  // add your custom rules here
  'rules': {
    // eslint rules
    'camelcase': ['error', {
      'properties': 'always'
    }],
    'function-paren-newline': ['error', 'consistent'],
    'id-match': ['error', '^(\\${0,1}[a-z]+[a-zA-Z_]*||[A-Z_0-9]+||[pk][12])$', {
      'onlyDeclarations': true,
    }],
    'max-len': ['error', {
      'code': 140,
      'ignoreTrailingComments': true,
      'ignoreStrings': true,
      'ignoreTemplateLiterals': true,
      'ignoreUrls': true,
      'ignoreComments': true
    }],
    'no-underscore-dangle': 0,
    'no-return-assign': 0,
    'object-curly-newline': ['error', {
      'consistent': true
    }],
    'one-var': ['error', {
      'initialized': 'never',
    }],
    'one-var-declaration-per-line': ['error', 'initializations'],
    'prefer-destructuring': 0,
    // allow debugger during development
    'no-debugger': isDev ? 0 : 2,
    'no-console': isDev ? 0 : 1,
    'no-unused-vars': isDev ? 0 : 1,

    // don't require .vue extension when importing
    'import/extensions': ['error', 'always', {
      'js': 'never',
      'vue': 'never'
    }],
    // allow optionalDependencies
    'import/no-extraneous-dependencies': ['error', {
      'optionalDependencies': ['test/unit/index.js']
    }],
    // allow single export
    'import/prefer-default-export': 'off',

    // vue lint configs
    'vue/attribute-hyphenation': ['error', 'always'],
    'vue/html-end-tags': 'error',
    'vue/html-indent': ['error', 2, {
      'attribute': 1,
      'closeBracket': 0,
      'ignores': []
    }],
    'vue/html-quotes': ['error', 'double'],
    'vue/html-self-closing': ['error', {
      'html': {
        'normal': 'never',
        'void': 'never',
        'component': 'never'
      },
      'svg': 'always',
      'math': 'always',
    }],
    'vue/max-attributes-per-line': [2, {
      'singleline': 10,
      'multiline': {
        'max': 2,
        'allowFirstLine': false
      },
    }],
    'vue/mustache-interpolation-spacing': ['error', 'always'],
    'vue/name-property-casing': ['error', 'kebab-case'],
    'vue/no-async-in-computed-properties': 'error',
    'vue/no-confusing-v-for-v-if': 'error',
    'vue/no-dupe-keys': 'error',
    'vue/no-duplicate-attributes': ['error', {
      allowCoexistClass:  true,
      allowCoexistStyle: true,
    }],
    'vue/no-multi-spaces': 'error',
    'vue/no-parsing-error': 'error',
    'vue/no-reserved-keys': ['error', {
      'reserved': ['$el', '$nextTick', '$route', '$router', 'asyncData'],
      'groups': [],
    }],
    'vue/no-shared-component-data': 'error',
    'vue/no-side-effects-in-computed-properties': 'error',
    'vue/no-template-key': 'error',
    'vue/no-textarea-mustache': 'error',
    // 'vue/order-in-components': ['error', {
    //   'order': [
    //     ['name', 'delimiters', 'functional', 'model'],
    //     ['components', 'directives', 'filters'],
    //     ['parent', 'mixins', 'extends', 'provide', 'inject'],
    //     'el',
    //     'template',
    //     'props',
    //     'propsData',
    //     'data',
    //     'computed',
    //     'watch',
    //     'asyncData',
    //     'onWechatReady',
    //     'LIFECYCLE_HOOKS',
    //     'methods',
    //     'render',
    //     'renderError'
    //   ],
    // }],
    'vue/require-component-is': 'error',
    'vue/require-default-prop': 'error',
    'vue/require-prop-types': 'error',
    'vue/require-render-return': 'error',
    'vue/require-v-for-key': 'error',
    'vue/require-valid-default-prop': 'error',
    'vue/return-in-computed-property': 'error',
    'vue/this-in-template': ['error', 'never'],
    'vue/v-bind-style': ['error', 'shorthand'],
    'vue/v-on-style': ['error', 'shorthand'],
    'vue/valid-template-root': 'error',
    'vue/valid-v-bind': 'error',
    'vue/valid-v-cloak': 'error',
    'vue/valid-v-else-if': 'error',
    'vue/valid-v-else': 'error',
    'vue/valid-v-for': 'error',
    'vue/valid-v-html': 'error',
    'vue/valid-v-if': 'error',
    'vue/valid-v-model': 'error',
    'vue/valid-v-on': 'error',
    'vue/valid-v-once': 'error',
    'vue/valid-v-pre': 'error',
    'vue/valid-v-show': 'error',
    'vue/valid-v-text': 'error',
  }
}

```

**What did you do? Please include the actual source code causing the issue.**

```vue
            <div class="pull-down-touching-tip__text">
              {{ scope.distance < fireDistance ? '继续下拉刷新' : '松手刷新' }}
            </div>
```

**What did you expect to happen?**

No error.

**What actually happened? Please include the actual, raw output from ESLint.**
![image](https://user-images.githubusercontent.com/1808990/35603738-a5fea0d8-0679-11e8-98ae-6a62c8b1c052.png)

**Comment by mysticatea** [comment]
Thank you for the report.

This is a correct error as according to HTML spec: 

- [invalid-first-character-of-tag-name error](https://html.spec.whatwg.org/multipage/parsing.html#parse-error-invalid-first-character-of-tag-name).
- [Tag open state](https://html.spec.whatwg.org/multipage/parsing.html#tag-open-state)'s `Anything else` step.

To make valid HTML, you should use `&lt;` instead of the `<`.
Or you can ignore the error by the option of `vue/no-parsing-error` rule: https://github.com/vuejs/eslint-plugin-vue/blob/master/docs/rules/no-parsing-error.md#wrench-options

**Comment by tinymins** [comment]
No this is NOT html, it's in `{{}}`, which means it is a inline JavaScript expression.

**Comment by mysticatea** [comment]
It's still in HTML.

**Comment by tinymins** [comment]
Thanks for reply. So what should I do to compare two numbers and use conditional operator correctly in vue template?

**Comment by tinymins** [comment]
The expression `{{ scope.distance < fireDistance ? 'Pull to refresh' : 'Release to refresh' }}` in template, will be rendered into `Pull to refresh` or `Release to refresh` due to `scope.distance` and  `fireDistance`. There is no spec in the result. I can't understand why lint report an error here. 😕

**Comment by mysticatea** [comment]
To be *valid HTML*, `{{ scope.distance &lt; fireDistance ? 'Pull to refresh' : 'Release to refresh' }}`.

But HTML spec allows `{{ scope.distance < fireDistance ? 'Pull to refresh' : 'Release to refresh' }}` as well due to well-defined error recovery logic. It's the reason that `vue/no-parsing-error` rule has the options to ignore HTML parse errors.

**Comment by tinymins** [comment]
Actually, `{{ scope.distance &lt; fireDistance ? 'Pull to refresh' : 'Release to refresh' }}` can not pass the compiler.
![image](https://user-images.githubusercontent.com/1808990/35608933-58fe1a80-0696-11e8-970d-5e2e8fbf7ee4.png)


It is really a great feature to detect specs in real HTML, and I need this very much. Some members in my team are always careless about use `&lt`. They think 'OH IT WORKS!', which makes me feel very annoying. That's why I do not want to disable it globally.

**Comment by mysticatea** [comment]
Oh? `{{ scope.distance &lt; fireDistance ? 'Pull to refresh' : 'Release to refresh' }}` works fine in my environment. It's strange.

**Comment by tinymins** [comment]
Full file `index.vue`:
```
<template>
  <div>A is {{ a > b ? 'larger than' : 'smaller or equals with' }} B</div>
</template>

<script>
export default {
  data() {
    return {
      a: Math.random(),
      b: Math.random(),
    };
  },
};
</script>
```
You can try this code.

**Comment by tinymins** [comment]
![image](https://user-images.githubusercontent.com/1808990/35611260-6e4d42cc-069f-11e8-93ad-3bd4cea1cc11.png)
Here's the expected result.

**Comment by mysticatea** [comment]
I know that HTML syntax is very tolerant. Browsers/Parsers work fine even if HTML parse error. But HTML parse errors still exist, so the rule reports it and has the options which are to ignore HTML parse errors.

OK, I opened an issue to add option which ignores HTML syntax errors in mustaches in `vue-eslint-parser` repo.

**Comment by mysticatea** [comment]
I re-thought about this, I decided that I don't add this feature.

It messes the code easily by removing a space. Inside of mustaches is still HTML.

```vue
<!-- Oops, this is not a mustache, there is `<fireDistance>` tag. -->
{{ scope.distance <fireDistance ? 'Pull to refresh' : 'Release to refresh' }}
```

I'd like to recommend to use `&lt;` (note it requires semicolon) for `<`.

Thank you.

**Comment by tinymins** [comment]
Thanks, I'll use &lt; inside the mustache.

**Comment by mysticatea** [comment]
> By that logic, anything inside of a <script> tag is HTML.

Exactly. So the following code is syntax error since the script ends at the first `</script>`. HTML spec allows `<` in script elements.

```html
<script>
    document.body.innerHTML += "</script>"
</script>
```

As same, `{{ a<b ? 1 : 2}}` is syntax error because `<b>` tag exists there. This is just a fact. I cannot change it.

**Comment by Nfinished** [comment]
So wouldn't it follow that `{{` ends at the first `}}`?

**Comment by mysticatea** [comment]
@Nfinished https://jsfiddle.net/67vmen98/

**Comment by Nfinished** [comment]
Not quite what I'm talking about.

https://jsfiddle.net/krvftv74/2/

It's kind of funny that my example is throwing HTML linter errors, but the way I see it the vue linter plugin should know better.

**Comment by ERPedersen** [comment]
> interpolation
> ɪntəːpəˈleɪʃ(ə)n/
> "the insertion of something of a different nature into something else".

The code inside the interpolation tags are **not** HTML, and therefore shouldn't be treated as such. This is ridiculous

**Comment by steffanhalv** [comment]
@mysticatea Please fix this, its really annoying :(

**Comment by liuyanjie** [comment]
Same problem, use `>`instead of `<`

```vue
<a v-for="item in items" :key="item._id">
  {{ (item.name.length > 6) ? item.name.slice(0,6) + '...' : item.name }}
</a>
```

![image](https://user-images.githubusercontent.com/5998826/46244842-b254e480-c417-11e8-84f5-75c6b3fac7a8.png)

**Comment by Smilebags** [comment]
This is still a thing, and it is absurd.

The contents of {{ }} is _executed as a JavaScript expression_. It isn't HTML by any stretch of the imagination. Having to escape html reserved characters is very unnatural in what is essentially a script.

Let JavaScript expressions be expressions. Fix this bug.

**Comment by tmcdos** [comment]
I completely agree with @Smilebags - Vue template is something more than a pure HTML and the linter should obey this fact. Everything inside a mustache and the mustache itself is and should be practically and theoretically invisible for the HTML parser.

**Comment by thejcannon** [comment]
I think we can all agree there are no HTML tags in `{{ a < b ? 1 : 2 }}`. 
So it follows that seeing an "invalid-first-character-of-__tag__-name" is a mistake.

Either the warning is valid or it isn't. And seeing how there aren't any tags in that statement, the warning is not valid.

I think we all understand the situation that leads to the warning being generated today, but the _why_ doesn't really change the _should_.

**Comment by fungiboletus** [comment]
Is the rule `invalid-first-character-of-tag-name` useful  ? Perhaps it can be removed if it's too challenging to fix it.

**Comment by hoomersinpsom** [comment]
Quick hotfix to this 
with error:
![image](https://user-images.githubusercontent.com/1879104/74217812-44fdc880-4c87-11ea-9e14-c3bf4d0d69cd.png)
no error:
![image](https://user-images.githubusercontent.com/1879104/74217843-5a72f280-4c87-11ea-88d1-61baf607ac43.png)

the rule `invalid-first-character-of-tag-name` should not be parsed inside mustache tags, inside it is javascript not html, dont matter what is your justification for this, its a compiler warning mustache tags are not suposed to be rendered in browser, so it doesnt make sense

**Comment by petsgre** [comment]
when can fix it ?

**Comment by diego-lipinski-de-castro** [comment]
LOL, still a problem

**Comment by emilio-ab** [comment]
still a problem

**Comment by helgard-mez** [comment]
Can we please get someone to re-evaluate this ticket? Based on the responses from @mysticatea it's quite evident that they either have a fundamental misunderstanding of how string interpolation works in Vue, or the purpose behind string interpolation in Vue.

**Comment by tmcdos** [comment]
Indeed, the problem is in [vue-eslint-parser](https://github.com/mysticatea/vue-eslint-parser) and not in **eslint-plugin-vue**. Here is the patch to fix the problem 
[vue-eslint-parser v7.0.0.patch.txt](https://github.com/vuejs/eslint-plugin-vue/files/4527792/vue-eslint-parser.v7.0.0.patch.txt)

**Comment by ota-meshi** [comment]
You can use the following options:

```json5
{
  // ...
  "rules": {
    // ...
    "vue/no-parsing-error": ["error", {"invalid-first-character-of-tag-name": false}]
    // ...
  }
  // ...
}
```


You can check with [DEMO].

[Demo]: https://mysticatea.github.io/vue-eslint-demo/#eJyNV9tunDoU/RXEYxUrR+cxGvUn2reTPhjYgBtjE9vMDI3y72fZBmNgIlVRlXjtjfdtedn9KGvdUPlSXhwNo+SOvr+qorg04vr946MQxeVS/FN8fl6ePfKqLs+Z38XWRowOfz1/K8hKoVxxnehZaTZyY4XqGBmjzUvxX/j9VHy8lkJduRQNa4WxjtU9N7x2ZJhumeMdU3yg1/KlaLm09Pnr27OPucQpn0ozSbLlywfSVtaZqXbaMDuNZMqXf5/KFqtGGKqd0CogHTm/uyE3mYggu5pbYg3VEsG9p00Gya1l+Ce6zVkPqIaYoo79IaM3XDUnV2W1pHzpHrg4rlz4XKQ0o8UZLZFqR/cVbKiaum6pLqwlOWJXviEonnHT2R0QCxloqMjsLW8054AUNYYZOrKiGLCbs8mEvfbGkfuuqh2YFvdDyXR3hrNKozNc+Uhub7E0iBXB2KXrjZ66PkGTqg87dlJXXB5AoRSSfTTVlXKhs2NCDdYTnNmtF47syOvUgkHcqWEBAh0wZ8ertJ2iG7PzUGm5Irr6jbqkTC66djxZDcWkaAOQx7L7ilmS7aEg6w+Rn63h28hcLyyrCESnjPcwTKqhdlugUJwCFDFM0gkcTdpshjimW8kMsrwlHEmFIuYDCt7znKeTmiz2lbyireAFBC83CCURSEi25mMMhT4hKWZ7bVyPtgbQ0PuEI8tmQTIi+JIJq3gMGWfn5pF0rM9LTGgKq8DQN3KhlVCb3KpvZ9Q5I6oJbO/nsSe1VbWzWqZNs/TVGyqp67fTZj40JuDmpYseq6FdMp2kgOhh4KzhqpN7kKAAi1Bdd5ZRK2/zMgjeslVt/bHJw9O7/0nL3g0S51QHzV27Ap6muX/pdKwrOFFg/CIpCRVgmHJ77H3Sjg5+gcpLnGT5be+eEXajiEchRqcMBu7qnm2taIWk0I/MAxKzjQuHgO0KTZRnsRpJod9eYENvD30ZJihy3ft+Q9RGLQMxTnmFkYxGI5yXx/08wHduZwgVZuZTn/zRW5xF1iD4rUKI48qRaG5Chu0UxnNl/iq7MtHm9r2Ab+Ai4ltPcnvoRq42j67o3GDIkrki/2MkGBAgqApqdfyeGy3uCxi2uTXc8Z2DaIhR24Lz9i/atIov7gKwERoBwffqGL5NleZfpLOCvB/iSLHRt73p7pA2ZysHcuOiaaki+8C4Y3MUvTA1dhOuj1PMzdfAyIQselhPRp7PQVChtVGHBHzLokYceLiK6TYGYU/GhXmh8WdjUCVczSNE+mQNgb0Ws+wB9rXXObaBtuzfY7k10j6fX7KEW+CLzP1ejyi1bRNfkEcJ862T9PdKETiJTVpxZ3q0B3xS3Mw7PFzWmYonQ6QJRIbueGMl+MoqZHi4VYKffwQFVfLPjL3p4B1vy5XyRmt3sMUgJxCCzd9OKK54ykUox09gTvYV2/F9BR9sOOC/IWfH7H7eoPoceTRnDK+M2wn0Bx7g51MZqfBD/IlfxuXP2b9UyvgWfCrDAww1lXgdGqLy839NfI6A

**Comment by quannt** [comment]
This is broken if you are using `@babel/plugin-proposal-optional-chaining` too.

```
<span>
  user?.name
</span>
```
throws an error.

**Comment by hoomersinpsom** [comment]
@mysticatea  can you reopen this issue? It's affecting other cases which will bring back the same discussion again

**Comment by mysticatea** [comment]
No. the inside of `{{ }}` is HTML in Vue.js 2.x. That's the reason that I rewrote HTML parser for Vue.js 3 from scratch.

This is just a fact. I don't think the reopen makes sense.

In Vue.js 3 support, this behavior will change, I think. Because the inside of `{{ }}` in Vue.js 3 is not HTML.

**Comment by stevebaros** [comment]
Go in packages.json and add this code
`rules: {
    "vue/no-parsing-error": [2, {
      "invalid-first-character-of-tag-name": false
    }]
  }`

**Comment by tmcdos** [comment]
> Go in packages.json and add this code

It is not `packages.json` but `.eslintrc.js`

**Comment by bumprat** [comment]
use `v-text` ( or `v-html` )

```html
<!-- will pass -->
<div v-text="a < b ? 'c' : 'd'"></div>
```
```html
<!-- will fail -->
<div>{{ a < b ? 'c' : 'd' }}</div>
```

**Comment by tmcdos** [comment]
@bumprat Are you sure your `v-text` will be compiled as expression rather than plain text ?

**Comment by bumprat** [comment]
> @bumprat Are you sure your `v-text` will be compiled as expression rather than plain text ?

check [v-text documentation](https://012.vuejs.org/api/directives.html#v-text)
> v-text
> Updates the element’s textContent.
> 
> Internally, {{ Mustache }} interpolations are also compiled as a v-text directive on a textNode.

**Comment by eithed** [comment]
> No. the inside of {{ }} is HTML in Vue.js 2.x. That's the reason that I rewrote HTML parser for Vue.js 3 from scratch.
>
> This is just a fact. I don't think the reopen makes sense.
>
> In Vue.js 3 support, this behavior will change, I think. Because the inside of {{ }} in Vue.js 3 is not HTML.

@mysticatea Just run into the issue with Vue.js 3. Is it possible to reopen the issue?

**Comment by ronny1020** [comment]
same problem here.

It can be fixed by add "vue/no-parsing-error": [2, { "invalid-first-character-of-tag-name": false }]" in .eslintrc.js .
However, I hope a better way to fix it.

**Comment by altin-at-casengo** [comment]
> No. the inside of `{{ }}` is HTML in Vue.js 2.x. That's the reason that I rewrote HTML parser for Vue.js 3 from scratch.
> 
> This is just a fact. I don't think the reopen makes sense.
> 
> In Vue.js 3 support, this behavior will change, I think. Because the inside of `{{ }}` in Vue.js 3 is not HTML.

Im using Vue 3 and I still get this error.

**Comment by trev-dev** [comment]
All other arguments about syntax aside, if JavaScript expressions (be them ternary or otherwise) are supported and encouraged by VueJS, it should probably not throw an error in a VueJS linter. `alt;` is not a JavaScript operator and anyone's opinion to the contrary is diverging standard.

**Comment by ysya** [comment]
still have this issue with Vue3 in 2021. I think this issue should be open.

**Comment by ota-meshi** [comment]
You need to set `parserOptions.vueFeatures.interpolationAsNonHTML` to true.

https://github.com/vuejs/vue-eslint-parser#parseroptionsvuefeaturesinterpolationasnonhtml
https://github.com/vuejs/vue-eslint-parser/pull/88

**Comment by manhhailua** [comment]
Still a problem in 09/2021.
My current solution is to use a method:
```
methods: {
    displayBigNumber(num: number): string {
      return num >= 10000 ? "9999+" : toString(num);
    },
}
```
And then:
```
<div>{{ displayBigNumber(10000) }}</div>
```
